server/svix-server/src/v1/endpoints/endpoint/crud.rs

```diff
@@ -47,156 +47,172 @@ pub(super) async fn list_endpoints(
+    let (endp, metadata) = {
+        let txn = ctx!(db.begin().await)?;
+        let endp = ctx!(endp.update(&txn).await)?;
+        let metadata = ctx!(metadata.upsert_or_delete(&txn).await)?;
+        ctx!(txn.commit().await)?;
+        (endp, metadata)
+    };
+
+    let app_uid = app.uid;
+    op_webhooks
+        .send_operational_webhook(
+            &app.org_id,
+            OperationalWebhook::EndpointUpdated(EndpointEvent::new(app_uid.as_ref(), &endp)),
+        )
+        .await?;
+
+    Ok((endp, metadata))
 }
 
 pub(super) async fn update_endpoint(
     Extension(ref db): Extension<DatabaseConnection>,
-    Extension(cfg): Extension<Configuration>,
-    Extension(op_webhooks): Extension<OperationalWebhookSender>,
+    Extension(ref cfg): Extension<Configuration>,
+    Extension(ref op_webhooks): Extension<OperationalWebhookSender>,
     Path((_app_id, endp_id)): Path<(ApplicationIdOrUid, EndpointIdOrUid)>,
     permissions::Application { app }: permissions::Application,
     ValidatedJson(mut data): ValidatedJson<EndpointIn>,
 ) -> Result<(StatusCode, Json<EndpointOut>)> {
-    let endp = ctx!(
-        endpoint::Entity::secure_find_by_id_or_uid(app.id.clone(), endp_id)
-            .one(db)
-            .await
-    )?;
-
     if let Some(ref event_types_ids) = data.event_types_ids {
         validate_event_types(db, event_types_ids, &app.org_id).await?;
     }
     validate_endpoint_url(&data.url, cfg.endpoint_https_only)?;
 
-    let app_uid = app.uid;
-
-    match endp {
-        Some(endp) => {
-            let mut endp: endpoint::ActiveModel = endp.into();
-            data.update_model(&mut endp);
-            let ret = ctx!(endp.update(db).await)?;
+    let models =
+        ctx!(endpoint::ActiveModel::fetch_with_metadata(db, app.id.clone(), endp_id).await)?;
 
-            op_webhooks
-                .send_operational_webhook(
-                    &app.org_id,
-                    OperationalWebhook::EndpointUpdated(EndpointEvent::new(app_uid.as_ref(), &ret)),
-                )
-                .await?;
-
-            Ok((StatusCode::OK, Json(ret.into())))
-        }
-        None => {
-            let ret = ctx!(
-                endpoint::ActiveModel {
-                    app_id: Set(app.id),
-                    key: Set(
-                        data.key_take_or_generate(&cfg.encryption, &cfg.default_signature_type)?
-                    ),
-                    ..data.into()
-                }
-                .insert(db)
-                .await
-            )?;
-
-            op_webhooks
-                .send_operational_webhook(
-                    &app.org_id,
-                    OperationalWebhook::EndpointCreated(EndpointEvent::new(app_uid.as_ref(), &ret)),
-                )
-                .await?;
-
-            Ok((StatusCode::CREATED, Json(ret.into())))
-        }
+    if let Some((mut endp, mut metadata)) = models {
+        metadata.data = Set(mem::take(&mut data.metadata));
```

Comment:
mem::take is scary, why do we need it?

Reply:
So all it does is just replace the value with it's `default()` (which in this case is just an empty map), and return the old value. It's not `unsafe` - I definitely wouldn't have used it if it was.

The advantage is just that it saves a `clone`, which for `metadata` seems worthwhile.